backend/src/main/scala/bloop/reporter/Reporter.scala

```diff
@@ -38,7 +40,10 @@ abstract class Reporter(
 
   private var _nextID = 1
   private def nextID(): Int = { val id = _nextID; _nextID += 1; id }
-  override def reset(): Unit = {
+  // FIXME: Adding a synchronized here is clearly not the proper fix, but it seems to help preventing random failures of the test. As
+  //        I'd like to see the test suite succeed on the CI, I've decided to include it in the current PR, even though we will of course
+  //        remove this before merging it.
+  override def reset(): Unit = synchronized {
```

Comment:
I wonder why `reset` is being called concurrently while compiling a project. I would expect Hydra doesn't call it unless all threads are done compiling a project. I'm concerned the fact we need `synchronized` might be making some other parts in Bloop not work well with Hydra, what do you think @dotta?

Reply:
You are correct that `reset` is not called concurrently. Maybe the race condition I was seeing was an artifact of the tests, but unfortunately I'm unable to reproduce the exception (I believe the exception I was seeing was an `ArrayIndexOutOfBoundException` with index -1).

Regardless, since errors may be reported concurrently, we still need to make the underlying `_problems` data structure thread-safe. That's exactly what I'm doing and the `synchronized` will disappear.

Reply:
Jorge, the call to `reporter.reset` is done in [`Global`](https://github.com/scala/scala/blob/2.13.x/src/compiler/scala/tools/nsc/Global.scala#L1493). That's why the race condition with the `Reporter._problems` could surface when compilation is carried out with  Hydra.

Reply:
Right, so it is indeed being called concurrently when the compiler is preparing the compilation units. Any chance you can change this in Hydra?

Reply:
Why would this need to be changed in Hydra and not in the vanilla Scala compiler too? The approach I'm taking is to make the Reporters implementation thread-safe, as we discussed.

Reply:
Even if you make the implementation thread-safe, could we not have a race condition here? If one hydra compile run finishes early before other compile runs and calls `reset` then the problems diagnosed by the other runs could be disappeared, no?

Reply:
`reset` is not called by Hydra. And because the current implementation works when used with the vanilla Scala compiler, there should be no reason for it to not work with Hydra.

Reply:
To clarify, `Reporter.reset` is called in Global as mentioned earlier, but this call happens _before_ the sources are compiled. So, while it's true there is a chance of a race-condition, empirical evidence from our experience with Hydra is that this call to `Reporter.reset` is generally executed across all Hydra threads before a problem is reported.

Reply:
> So, while it's true there is a chance of a race-condition, empirical evidence from our experience with Hydra is that this call to Reporter.reset

Sure, so long as we're explicit about this and we acknowledge that this race condition could happen in the future or could be reported by users, I'm fie with it :smile:

Reply:
Sounds good. I've actually filed a ticket on our side as I think this is should be easy to fix (and I think we can have a fix included in the next Hydra release). I'm going to marke this discussion resolved as there is no further action to take here.